cfgorch/cfgorch.cpp

```diff
@@ -0,0 +1,176 @@
+    }
+    return selectables;
+}
+
+bool CfgOrch::hasSelectable(TableConsumable *selectable) const
+{
+    for(auto it : m_consumerMap) {
+        if (it.second.m_consumer == selectable) {
+            return true;
+        }
+    }
+    return false;
+}
+
+bool CfgOrch::syncCfgDB(string tableName, Table &tableConsumer)
+{
+    SWSS_LOG_ENTER();
+
+    auto consumer_it = m_consumerMap.find(tableName);
+    if (consumer_it == m_consumerMap.end())
+    {
+        SWSS_LOG_ERROR("Unrecognized tableName:%s\n", tableName.c_str());
+        return false;
+    }
+    Consumer& consumer = consumer_it->second;
+
+    vector<KeyOpFieldsValuesTuple> tuples;
+
+    tableConsumer.getTableContent(tuples);
+    for (auto tuple : tuples)
+    {
+        string key = kfvKey(tuple);
+        /* Directly put it into consumer.m_toSync map */
+        if (consumer.m_toSync.find(key) == consumer.m_toSync.end())
+        {
+            consumer.m_toSync[key] = make_tuple(key, SET_COMMAND, kfvFieldsValues(tuple));
+            SWSS_LOG_DEBUG("%s", (dumpTuple(consumer, tuple)).c_str());
+        }
+        /*
+         * Syncing from DB directly, don't expect duplicate keys.
+         * Or there is pending task from consumber state pipe, in this case just skip it.
+         */
+        else
+        {
+            SWSS_LOG_WARN("Duplicate key %s found in tableName:%s\n", key.c_str(), tableName.c_str());
+            continue;
+        }
+        doTask(consumer);
+    }
+    return true;
+}
+
+bool CfgOrch::execute(string tableName)
+{
+    SWSS_LOG_ENTER();
+
+    auto consumer_it = m_consumerMap.find(tableName);
+    if (consumer_it == m_consumerMap.end())
+    {
+        SWSS_LOG_ERROR("Unrecognized tableName:%s\n", tableName.c_str());
+        return false;
+    }
+    Consumer& consumer = consumer_it->second;
+
+    int data_popped = 0;
+    while (1)
+    {
+        KeyOpFieldsValuesTuple new_data;
+        consumer.m_consumer->pop(new_data);
+
+        string key = kfvKey(new_data);
+        string op  = kfvOp(new_data);
+        /*
+         * Done with all new data. Or
+         * possible nothing popped, ie. the oparation is already merged with other operations
+         */
+        if (op.empty())
+        {
+            SWSS_LOG_DEBUG("Number of kfv data popped: %d\n", data_popped);
+            break;
+        }
+        data_popped++;
+        SWSS_LOG_DEBUG("%s", (dumpTuple(consumer, new_data)).c_str());
+
+        /* If a new task comes or if a DEL task comes, we directly put it into consumer.m_toSync map */
+        if (consumer.m_toSync.find(key) == consumer.m_toSync.end() || op == DEL_COMMAND)
+        {
+           consumer.m_toSync[key] = new_data;
+        }
+        /* If an old task is still there, we combine the old task with new task */
+        else
+        {
+            KeyOpFieldsValuesTuple existing_data = consumer.m_toSync[key];
+
+            auto new_values = kfvFieldsValues(new_data);
```

Comment:
>auto [](start = 12, length = 4)

Many places, you could use "auto &" to reduce copying cost.

Reply:
ok, will check how to sync with orch class too.